packages/flutter/lib/src/widgets/app.dart

```diff
@@ -23,6 +23,32 @@ import 'widget_inspector.dart';
 
 export 'dart:ui' show Locale;
 
+/// The signature of [WidgetsApp.onLocaleChanged] callback.
+///
+/// A `LocaleChangedCallback` is responsible for computing the locale of the app's
+/// [Localizations] object when the app starts and when user changes the default
+/// locale for the device.
+///
+/// The `oldLocale` is null when the app starts. The `newLocale` is the device's
+/// locale when the app started, or the device locale the user selected after
+/// the app was started. The `supportedLocales` parameter is just the value of
+/// [WidgetApp.supportedLocales].
+typedef Locale LocaleChangedCallback(Locale oldLocale, Locale newLocale, Iterable<Locale> supportedLocales);
+
+/// The default value of the [WidgetsApp.onLocaleChanged] callback.
+///
+/// If `supportedLocales` contains `newLocale` then return it. Otherwise return
+/// the first supported locale whose [Locale.languageCode] matches the new locale's
+/// language code. Failing that, return `oldLocale`, or the first supported
+/// locale if the old locale is null (because this app was just started).
```

Comment:
we should never return oldLocale. In fact I would be tempted to say we shouldn't even pass in oldLocale.

The reason we shouldn't use oldLocale is that it would mean the application will behave differently based on the order of locale changes, or based on whether it's restarted or not (restarting the app without changing the system locale shouldn't change the locale).

It should IMHO just default to the first locale in the list if it can't find a good one.

Reply:
OK, I've pitched the oldLocale parameter.